plugin/src/helpers/config.ts

```diff
@@ -114,64 +115,133 @@ export async function checkConfig({ utils, netlifyConfig }): Promise<void> {
+        path.posix.join(cacheDir, 'query-engine', '**'),
+        path.posix.join(cacheDir, 'page-ssr', '**'),
+        '!**/*.js.map',
+      ],
+      external_node_modules: ['msgpackr-extract'],
+      node_bundler: 'esbuild',
+    }
   }
 
-  netlifyConfig.functions.__ssr = { ...netlifyConfig.functions.__dsg }
+  if (neededFunctions.includes('SSR')) {
+    netlifyConfig.functions.__ssr = {
+      included_files: [
+        'public/404.html',
+        'public/500.html',
+        path.posix.join(cacheDir, 'data', '**'),
+        path.posix.join(cacheDir, 'query-engine', '**'),
+        path.posix.join(cacheDir, 'page-ssr', '**'),
+        '!**/*.js.map',
+      ],
+      external_node_modules: ['msgpackr-extract'],
+      node_bundler: 'esbuild',
+    }
+  }
   /* eslint-enable no-underscore-dangle, no-param-reassign */
 }
 
-export function shouldSkipFunctions(cacheDir: string): boolean {
-  if (
-    process.env.NETLIFY_SKIP_GATSBY_FUNCTIONS === 'true' ||
-    process.env.NETLIFY_SKIP_GATSBY_FUNCTIONS === '1'
-  ) {
-    console.log(
-      'Skipping Gatsby Functions and SSR/DSG support because the environment variable NETLIFY_SKIP_GATSBY_FUNCTIONS is set to true',
-    )
-    return true
-  }
+export async function getNeededFunctions(
+  cacheDir: string,
+): Promise<FunctionList> {
+  if (!existsSync(path.join(cacheDir, 'functions'))) return []
 
-  if (!existsSync(path.join(cacheDir, 'functions'))) {
-    console.log(
-      `Skipping Gatsby Functions and SSR/DSG support because the site's Gatsby version does not support them`,
-    )
-    return true
+  const neededFunctions = overrideNeededFunctions(
+    await readFunctionSkipFile(cacheDir),
+  )
+
+  const functionList = Object.keys(neededFunctions).filter(
+    (name) => neededFunctions[name] === true,
+  ) as FunctionList
+
+  if (functionList.length === 0) {
+    console.log('Skipping Gatsby Functions and SSR/DSG support')
+  } else {
+    console.log(`Enabling Gatsby ${functionList.join('/')} support`)
   }
 
-  const skipFile = path.join(cacheDir, '.nf-skip-gatsby-functions')
+  return functionList
+}
 
-  if (existsSync(skipFile)) {
-    console.log(
-      stripIndent`
-      Skipping Gatsby Functions and SSR/DSG support because gatsby-plugin-netlify reported that this site does not use them. 
-      If this is incorrect, remove the file "${skipFile}" and try again.`,
-    )
-    return true
+async function readFunctionSkipFile(cacheDir: string) {
+  try {
+    // read skip file from gatsby-plugin-netlify
+    return await fs.readJson(path.join(cacheDir, '.nf-skip-gatsby-functions'))
+  } catch (error) {
+    // missing skip file = all functions needed
+    // empty or invalid skip file = no functions needed
+    return error.code === 'ENOENT' ? { API: true, SSR: true, DSG: true } : {}
   }
+}
+
+// eslint-disable-next-line complexity
+function overrideNeededFunctions(neededFunctions) {
+  const skipAll =
+    process.env.NETLIFY_SKIP_GATSBY_FUNCTIONS === 'true' ||
```

Comment:
[sand] I wonder if these repeated checks could be abstracted into a helper that checks for true or 1